### PR TITLE
Caller specified CPI compute budget, invoke_signed_with_budget

### PR DESCRIPTION
**Pull request by jon-chuang**
#### Problem
Programs have no control over the compute limits of inner CPI instructions.

That makes them powerless to defend against long tail-end CU consumption from inner invocations.

Having more control in this area would allow programs to more confidently depend on the median or p75/p95 consumed CUs of inner CPI instructions.

If exceeding the local compute limit, the inner instruction will terminate early. One can leverage existing infrastructure (both in bpf interpreter loop, and in JIT branch checks) for early instruction termination.
#### Summary of Changes
A program, when making a CPI, ought to be able to limit inner instructions in the compute budget they consume. They would then be able to handle errors returned by invoke_signed_with_budget(..) -> ProgramResult appropriately.

To facilitate this, upon calling invoke_signed_with_budget, one should create a new inner compute meter. The inner compute meter's budget is the min of the outer's remaining budget and its caller specified budget. When the inner instruction is complete, the outer compute meter checks the total consumed, and subtracts that from its own compute meter.

#### Miscellaeneous
We don't need to make this more generic (over non-CPI) as with `sol_remaining_compute_units`, programs would already have control over within-program instructions (there's also no way to do the exact same thing - perform copy-on-write and commit on success as with CPI for within-program calls).

Note that the budget does not merely apply to bpf programs, but also whatever program that is called (for instance a system program). For what it's worth, I'm not sure if calls to say the native programs like `system_program` and `stake_program`  incur any cost at all - which is weird enough to be sure, and I'm sure can also be abused.

Fixes https://github.com/solana-labs/solana/issues/18243

### TODO:
- [x] Example program, which fails with `invoke_signed` but successfully handles a local computational budget exceeded error with `invoke_signed_with_budget`. Also show via `sol_log_compute_units` that the inner calls with `invoke_signed_with_budget` have a reduced budget :)
- [x] ~~Return status code that gets converted to a `ProgramError::Custom(0x0b9f_05c1)`~~
- [x] Piggyback on other syscall-related features
- [x] Add C version of syscall 
- [x] Add non "signed" version of syscall
- [x] Update docs with usage example
- [ ] Add example C program
- [x] Combine the Rust and C `call_with_budget` code blocks for maintainability.
- [x] Instead of utilising a status code, directly utilise a `ProgramError::ComputationalBudgetExceeded`
- [x] Return error changes should be feature-gated too
- [x] Also add API that allows rollbacks on arbitrary errors, `invoke_with_rollback` (https://github.com/solana-labs/solana/issues/18325)
- [x] Example program for `invoke_with_rollback`
- [x] tests in `message_processor` showing that per-PreAccount update stacks work as expected (in particular to facilitate rollbacks).
- [x] Rename underlying syscalls to `InvokeSignedWithOptions`
- [x] Reduce code bloat in API wrappers
- [x] Wrap `invoke_with_rollback` error type as `InvokeError`
- [ ] Split up PR into 1. MessageProcessor rollback capability, 2. Syscalls and APIs
- [ ] Improve rollback tests to show that previous state changes from lower invocations persist despite current invocation failing.
 
Note that I have not actually tested the C sdk code

### CODE REVIEW

**Comment by jon-chuang** [comment]
Output from locally run `cargo test-bpf` of the example program `cpi_budgeted_early_termination`:
```
[2021-06-27T08:50:15.204579900Z DEBUG solana_runtime::message_processor] Program outer11111111111111111111111111111111111111 invoke [1]
[2021-06-27T08:50:15.204902781Z DEBUG solana_runtime::message_processor] Program consumption: 199549 units remaining
[2021-06-27T08:50:15.217004922Z DEBUG solana_runtime::message_processor] Program 1nner11111111111111111111111111111111111111 invoke [2]
[2021-06-27T08:50:15.217172753Z DEBUG solana_runtime::message_processor] Program consumption: 23797 units remaining
[2021-06-27T08:50:15.219501856Z DEBUG solana_runtime::message_processor] Program consumption: 15295 units remaining
[2021-06-27T08:50:15.221737160Z DEBUG solana_runtime::message_processor] Program consumption: 6793 units remaining
[2021-06-27T08:50:15.223842487Z DEBUG solana_runtime::message_processor] Program 1nner11111111111111111111111111111111111111 consumed 24000 of 24000 compute units
[2021-06-27T08:50:15.223889561Z DEBUG solana_runtime::message_processor] Program failed to complete: exceeded maximum number of instructions allowed (24000) at instruction #208
[2021-06-27T08:50:15.223924412Z DEBUG solana_runtime::message_processor] Program 1nner11111111111111111111111111111111111111 failed: Program failed to complete
[2021-06-27T08:50:15.224072198Z DEBUG solana_runtime::message_processor] Program log: inner CPI failed to complete
[2021-06-27T08:50:15.224137990Z DEBUG solana_runtime::message_processor] Program consumption: 173993 units remaining
[2021-06-27T08:50:15.231331332Z DEBUG solana_runtime::message_processor] Program 1nner11111111111111111111111111111111111111 invoke [2]
[2021-06-27T08:50:15.231478001Z DEBUG solana_runtime::message_processor] Program consumption: 23797 units remaining
[2021-06-27T08:50:15.233461942Z DEBUG solana_runtime::message_processor] Program consumption: 15295 units remaining
[2021-06-27T08:50:15.235417318Z DEBUG solana_runtime::message_processor] Program consumption: 6793 units remaining
[2021-06-27T08:50:15.237079147Z DEBUG solana_runtime::message_processor] Program 1nner11111111111111111111111111111111111111 consumed 24000 of 24000 compute units
[2021-06-27T08:50:15.237123566Z DEBUG solana_runtime::message_processor] Program failed to complete: exceeded maximum number of instructions allowed (24000) at instruction #208
[2021-06-27T08:50:15.237158837Z DEBUG solana_runtime::message_processor] Program 1nner11111111111111111111111111111111111111 failed: Program failed to complete
[2021-06-27T08:50:15.237285042Z DEBUG solana_runtime::message_processor] Program log: inner CPI failed to complete
[2021-06-27T08:50:15.237340706Z DEBUG solana_runtime::message_processor] Program consumption: 148437 units remaining
[2021-06-27T08:50:15.243727229Z DEBUG solana_runtime::message_processor] Program 1nner11111111111111111111111111111111111111 invoke [2]
[2021-06-27T08:50:15.243879206Z DEBUG solana_runtime::message_processor] Program consumption: 23797 units remaining
[2021-06-27T08:50:15.245872227Z DEBUG solana_runtime::message_processor] Program consumption: 15295 units remaining
[2021-06-27T08:50:15.247872721Z DEBUG solana_runtime::message_processor] Program consumption: 6793 units remaining
[2021-06-27T08:50:15.249522816Z DEBUG solana_runtime::message_processor] Program 1nner11111111111111111111111111111111111111 consumed 24000 of 24000 compute units
[2021-06-27T08:50:15.249547191Z DEBUG solana_runtime::message_processor] Program failed to complete: exceeded maximum number of instructions allowed (24000) at instruction #208
[2021-06-27T08:50:15.249583859Z DEBUG solana_runtime::message_processor] Program 1nner11111111111111111111111111111111111111 failed: Program failed to complete
[2021-06-27T08:50:15.249792478Z DEBUG solana_runtime::message_processor] Program log: inner CPI failed to complete
[2021-06-27T08:50:15.249865323Z DEBUG solana_runtime::message_processor] Program consumption: 122881 units remaining
[2021-06-27T08:50:15.256083177Z DEBUG solana_runtime::message_processor] Program 1nner11111111111111111111111111111111111111 invoke [2]
[2021-06-27T08:50:15.256240602Z DEBUG solana_runtime::message_processor] Program consumption: 23797 units remaining
[2021-06-27T08:50:15.258262956Z DEBUG solana_runtime::message_processor] Program consumption: 15295 units remaining
[2021-06-27T08:50:15.260285870Z DEBUG solana_runtime::message_processor] Program consumption: 6793 units remaining
[2021-06-27T08:50:15.262108755Z DEBUG solana_runtime::message_processor] Program 1nner11111111111111111111111111111111111111 consumed 24000 of 24000 compute units
[2021-06-27T08:50:15.262219385Z DEBUG solana_runtime::message_processor] Program failed to complete: exceeded maximum number of instructions allowed (24000) at instruction #208
[2021-06-27T08:50:15.262256052Z DEBUG solana_runtime::message_processor] Program 1nner11111111111111111111111111111111111111 failed: Program failed to complete
[2021-06-27T08:50:15.262390639Z DEBUG solana_runtime::message_processor] Program log: inner CPI failed to complete
[2021-06-27T08:50:15.262448328Z DEBUG solana_runtime::message_processor] Program consumption: 97325 units remaining
[2021-06-27T08:50:15.268792597Z DEBUG solana_runtime::message_processor] Program 1nner11111111111111111111111111111111111111 invoke [2]
[2021-06-27T08:50:15.269018746Z DEBUG solana_runtime::message_processor] Program consumption: 23797 units remaining
[2021-06-27T08:50:15.271040193Z DEBUG solana_runtime::message_processor] Program consumption: 15295 units remaining
[2021-06-27T08:50:15.273061989Z DEBUG solana_runtime::message_processor] Program consumption: 6793 units remaining
[2021-06-27T08:50:15.274714878Z DEBUG solana_runtime::message_processor] Program 1nner11111111111111111111111111111111111111 consumed 24000 of 24000 compute units
[2021-06-27T08:50:15.274739113Z DEBUG solana_runtime::message_processor] Program failed to complete: exceeded maximum number of instructions allowed (24000) at instruction #208
[2021-06-27T08:50:15.274781927Z DEBUG solana_runtime::message_processor] Program 1nner11111111111111111111111111111111111111 failed: Program failed to complete
[2021-06-27T08:50:15.274939351Z DEBUG solana_runtime::message_processor] Program log: inner CPI failed to complete
[2021-06-27T08:50:15.275017854Z DEBUG solana_runtime::message_processor] Program consumption: 71769 units remaining
[2021-06-27T08:50:15.281483927Z DEBUG solana_runtime::message_processor] Program 1nner11111111111111111111111111111111111111 invoke [2]
[2021-06-27T08:50:15.281658044Z DEBUG solana_runtime::message_processor] Program consumption: 23797 units remaining
[2021-06-27T08:50:15.283752546Z DEBUG solana_runtime::message_processor] Program consumption: 15295 units remaining
[2021-06-27T08:50:15.285797600Z DEBUG solana_runtime::message_processor] Program consumption: 6793 units remaining
[2021-06-27T08:50:15.287446438Z DEBUG solana_runtime::message_processor] Program 1nner11111111111111111111111111111111111111 consumed 24000 of 24000 compute units
[2021-06-27T08:50:15.287473606Z DEBUG solana_runtime::message_processor] Program failed to complete: exceeded maximum number of instructions allowed (24000) at instruction #208
[2021-06-27T08:50:15.287516140Z DEBUG solana_runtime::message_processor] Program 1nner11111111111111111111111111111111111111 failed: Program failed to complete
[2021-06-27T08:50:15.287694797Z DEBUG solana_runtime::message_processor] Program log: inner CPI failed to complete
[2021-06-27T08:50:15.287753255Z DEBUG solana_runtime::message_processor] Program consumption: 46213 units remaining
[2021-06-27T08:50:15.294350283Z DEBUG solana_runtime::message_processor] Program 1nner11111111111111111111111111111111111111 invoke [2]
[2021-06-27T08:50:15.294507079Z DEBUG solana_runtime::message_processor] Program consumption: 23797 units remaining
[2021-06-27T08:50:15.296525243Z DEBUG solana_runtime::message_processor] Program consumption: 15295 units remaining
[2021-06-27T08:50:15.298562474Z DEBUG solana_runtime::message_processor] Program consumption: 6793 units remaining
[2021-06-27T08:50:15.300265510Z DEBUG solana_runtime::message_processor] Program 1nner11111111111111111111111111111111111111 consumed 24000 of 24000 compute units
[2021-06-27T08:50:15.300292749Z DEBUG solana_runtime::message_processor] Program failed to complete: exceeded maximum number of instructions allowed (24000) at instruction #208
[2021-06-27T08:50:15.300328927Z DEBUG solana_runtime::message_processor] Program 1nner11111111111111111111111111111111111111 failed: Program failed to complete
[2021-06-27T08:50:15.300483139Z DEBUG solana_runtime::message_processor] Program log: inner CPI failed to complete
[2021-06-27T08:50:15.300541876Z DEBUG solana_runtime::message_processor] Program consumption: 20658 units remaining
[2021-06-27T08:50:15.300611020Z DEBUG solana_rbpf::vm] BPF instructions executed (interp): 3551
[2021-06-27T08:50:15.300624150Z DEBUG solana_rbpf::vm] Max frame depth reached: 6
[2021-06-27T08:50:15.300662494Z DEBUG solana_runtime::message_processor] Program outer11111111111111111111111111111111111111 consumed 179447 of 200000 compute units
[2021-06-27T08:50:15.300790794Z DEBUG solana_runtime::message_processor] Program outer11111111111111111111111111111111111111 success
```

**Comment by jon-chuang** [comment]
@mvines @jackcmay ready for review. Need some clarification if a C program example is necessary.

**Comment by mvines** [comment]
I'm bearish on this feature, I don't see the clear need especially with #18227.  Without a clear need, by default I trend towards "no" because a smaller runtime is easier to maintain.

Would like to hear @jackcmay's thoughts though when he's back

**Comment by jon-chuang** [comment]
Hi @mvines , let me give an example use case:
1. I have a contract needing to conduct a swap on a DEX. It prefers to do so on the Raydium DEX. However, Raydium is known to sometimes have high compute cost (in the 100K-800Ks). I want to ensure that the transaction does not exceed its budget. I want to fall back on swapping on Serum directly if the Raydium instruction uses too many compute units. With https://github.com/solana-labs/solana/pull/18227, you are unable to achieve this. But with `invoke_with_budget` you can.

**Comment by jon-chuang** [comment]
Note that the implementation here plays well with https://github.com/solana-labs/solana/pull/18158 as the invoke context's compute meter is hot-swapped, and then swapped back, within the context of a single instruction.

In @jackcmay 's PR, the invoke context's compute meter is shared between instructions of a tx's message.

**Comment by mvines** [comment]
I feel like this could also be done off-chain. You could simulate the Raydium swap and if it's too expensive swap on Serum instead. I mean it's cool to do it on-chain but do you really need to.

**Comment by jon-chuang** [comment]
> I feel like this could also be done off-chain. You could simulate the Raydium swap and if it's too expensive swap on Serum instead. I mean it's cool to do it on-chain but do you really need to.

It's possible to try to submit the appropriate transaction with simulation. But the sands may shift by the time you submit your transaction. A small change in state can cause the compute cost of a CPI call to behave very differently during simulation and when executing. This makes it difficult to determine in advance which code path you'd best be taking.

Having it on-chain guarantees you can secure the swap, with fallback, regardless of how the sands have shifted.

**Comment by mvines** [comment]
Yeah that's fair.  WFM, if @jackcmay is down

**Comment by jon-chuang** [comment]
@jackcmay Need some advice on whether you'd like me to split this PR into two parts
1. the capabilities necessary for rollback requiring changes in `MessageProcessor` (described in https://github.com/solana-labs/solana/issues/18325) 
2. the new syscalls and APIs for invoking with budget (and rolling back on `ComputationalBudgetExceeded` and for rolling back on arbitrary errors

File: sdk/bpf/c/inc/solana_sdk.h [hunk 1]
```diff
@@ -606,6 +615,102 @@ static uint64_t sol_invoke(
...
+ */
+static uint64_t sol_invoke_signed_with_budget(
+    const SolInstruction *instruction,
+    int budget,
+    const SolAccountInfo *account_infos,
+    int account_infos_len,
+    const SolSignerSeeds *signers_seeds,
+    int signers_seeds_len
+) {
+  const SolInstructionWithOptions instruction_with_options = { (uint64_t)instruction, (uint64_t)budget, false };
```
Comment (jon-chuang) [inline sdk/bpf/c/inc/solana_sdk.h:680]: should be true, not false

**Comment by jon-chuang** [comment]
Here are the logs when running the new program example `invoke_with_rollback`:
```
[2021-07-03T17:30:10.247389730Z INFO  solana_program_test] "outer" program loaded as native code
[2021-07-03T17:30:10.247496800Z INFO  solana_program_test] "error" program loaded as native code
[2021-07-03T17:30:10.247536384Z INFO  solana_program_test] "inner" program loaded as native code
[2021-07-03T17:30:10.537568451Z DEBUG solana_runtime::message_processor] Program log: SyscallStubs: sol_log_compute_units() not available
[2021-07-03T17:30:10.537683947Z DEBUG solana_runtime::message_processor] Program error11111111111111111111111111111111111111 invoke [1]
[2021-07-03T17:30:10.537776270Z DEBUG solana_runtime::message_processor] Program log: invoke_with_rollback: error 1, rolled back
[2021-07-03T17:30:10.537845359Z DEBUG solana_runtime::message_processor] Program inner11111111111111111111111111111111111111 invoke [1]
[2021-07-03T17:30:10.537901144Z DEBUG solana_runtime::message_processor] Program log: SyscallStubs: sol_log_compute_units() not available
[2021-07-03T17:30:10.537926691Z DEBUG solana_runtime::message_processor] Program log: SyscallStubs: sol_log_compute_units() not available
[2021-07-03T17:30:10.537977837Z DEBUG solana_runtime::message_processor] Program inner11111111111111111111111111111111111111 success
[2021-07-03T17:30:10.538008975Z DEBUG solana_runtime::message_processor] Program log: SyscallStubs: sol_log_compute_units() not available
[2021-07-03T17:30:10.538073065Z DEBUG solana_runtime::message_processor] Program error11111111111111111111111111111111111111 invoke [1]
[2021-07-03T17:30:10.538135633Z DEBUG solana_runtime::message_processor] Program log: invoke_with_rollback: error 1, rolled back
[2021-07-03T17:30:10.538202858Z DEBUG solana_runtime::message_processor] Program inner11111111111111111111111111111111111111 invoke [1]
[2021-07-03T17:30:10.538257130Z DEBUG solana_runtime::message_processor] Program log: SyscallStubs: sol_log_compute_units() not available
[2021-07-03T17:30:10.538282608Z DEBUG solana_runtime::message_processor] Program log: SyscallStubs: sol_log_compute_units() not available
[2021-07-03T17:30:10.538333002Z DEBUG solana_runtime::message_processor] Program inner11111111111111111111111111111111111111 success
[2021-07-03T17:30:10.538353280Z DEBUG solana_runtime::message_processor] Program log: SyscallStubs: sol_log_compute_units() not available
[2021-07-03T17:30:10.538415266Z DEBUG solana_runtime::message_processor] Program error11111111111111111111111111111111111111 invoke [1]
[2021-07-03T17:30:10.538477122Z DEBUG solana_runtime::message_processor] Program log: invoke_with_rollback: error 1, rolled back
[2021-07-03T17:30:10.538544057Z DEBUG solana_runtime::message_processor] Program inner11111111111111111111111111111111111111 invoke [1]
[2021-07-03T17:30:10.538598228Z DEBUG solana_runtime::message_processor] Program log: SyscallStubs: sol_log_compute_units() not available
[2021-07-03T17:30:10.538623125Z DEBUG solana_runtime::message_processor] Program log: SyscallStubs: sol_log_compute_units() not available
[2021-07-03T17:30:10.538673088Z DEBUG solana_runtime::message_processor] Program inner11111111111111111111111111111111111111 success
[2021-07-03T17:30:10.538692775Z DEBUG solana_runtime::message_processor] Program log: SyscallStubs: sol_log_compute_units() not available
[2021-07-03T17:30:10.538754300Z DEBUG solana_runtime::message_processor] Program error11111111111111111111111111111111111111 invoke [1]
[2021-07-03T17:30:10.538815996Z DEBUG solana_runtime::message_processor] Program log: invoke_with_rollback: error 1, rolled back
[2021-07-03T17:30:10.538883332Z DEBUG solana_runtime::message_processor] Program inner11111111111111111111111111111111111111 invoke [1]
[2021-07-03T17:30:10.538940399Z DEBUG solana_runtime::message_processor] Program log: SyscallStubs: sol_log_compute_units() not available
[2021-07-03T17:30:10.538978360Z DEBUG solana_runtime::message_processor] Program log: SyscallStubs: sol_log_compute_units() not available
[2021-07-03T17:30:10.539082354Z DEBUG solana_runtime::message_processor] Program inner11111111111111111111111111111111111111 success
[2021-07-03T17:30:10.539123081Z DEBUG solana_runtime::message_processor] Program log: SyscallStubs: sol_log_compute_units() not available
[2021-07-03T17:30:10.539250229Z DEBUG solana_runtime::message_processor] Program error11111111111111111111111111111111111111 invoke [1]
[2021-07-03T17:30:10.539425787Z DEBUG solana_runtime::message_processor] Program log: invoke_with_rollback: error 1, rolled back
[2021-07-03T17:30:10.539574345Z DEBUG solana_runtime::message_processor] Program inner11111111111111111111111111111111111111 invoke [1]
[2021-07-03T17:30:10.539729365Z DEBUG solana_runtime::message_processor] Program log: SyscallStubs: sol_log_compute_units() not available
[2021-07-03T17:30:10.539771013Z DEBUG solana_runtime::message_processor] Program log: SyscallStubs: sol_log_compute_units() not available
[2021-07-03T17:30:10.539847867Z DEBUG solana_runtime::message_processor] Program inner11111111111111111111111111111111111111 success
[2021-07-03T17:30:10.539892681Z DEBUG solana_runtime::message_processor] Program log: SyscallStubs: sol_log_compute_units() not available
[2021-07-03T17:30:10.540019418Z DEBUG solana_runtime::message_processor] Program error11111111111111111111111111111111111111 invoke [1]
[2021-07-03T17:30:10.540153118Z DEBUG solana_runtime::message_processor] Program log: invoke_with_rollback: error 1, rolled back
[2021-07-03T17:30:10.540222979Z DEBUG solana_runtime::message_processor] Program inner11111111111111111111111111111111111111 invoke [1]
[2021-07-03T17:30:10.540279505Z DEBUG solana_runtime::message_processor] Program log: SyscallStubs: sol_log_compute_units() not available
[2021-07-03T17:30:10.540305714Z DEBUG solana_runtime::message_processor] Program log: SyscallStubs: sol_log_compute_units() not available
[2021-07-03T17:30:10.540357220Z DEBUG solana_runtime::message_processor] Program inner11111111111111111111111111111111111111 success
[2021-07-03T17:30:10.540377518Z DEBUG solana_runtime::message_processor] Program log: SyscallStubs: sol_log_compute_units() not available
[2021-07-03T17:30:10.540439234Z DEBUG solana_runtime::message_processor] Program error11111111111111111111111111111111111111 invoke [1]
[2021-07-03T17:30:10.540502142Z DEBUG solana_runtime::message_processor] Program log: invoke_with_rollback: error 1, rolled back
[2021-07-03T17:30:10.540569889Z DEBUG solana_runtime::message_processor] Program inner11111111111111111111111111111111111111 invoke [1]
[2021-07-03T17:30:10.540625052Z DEBUG solana_runtime::message_processor] Program log: SyscallStubs: sol_log_compute_units() not available
[2021-07-03T17:30:10.540650189Z DEBUG solana_runtime::message_processor] Program log: SyscallStubs: sol_log_compute_units() not available
[2021-07-03T17:30:10.540700333Z DEBUG solana_runtime::message_processor] Program inner11111111111111111111111111111111111111 success
[2021-07-03T17:30:10.540720671Z DEBUG solana_runtime::message_processor] Program log: SyscallStubs: sol_log_compute_units() not available

```

**Pull request closed without merging**